File: board/memory.js
import MersenneTwister from 'mersennetwister';
import { range, concatLists } from './util.js';

// Create the neighborhood memory map and the lookup tables that live at 0xE000
const taxicab = (vec) => Math.abs(vec[0]) + Math.abs(vec[1]);
const maxDelta = (vec) => Math.max(Math.abs(vec[0]), Math.abs(vec[1]));
const posAngle = (angle) => angle < 0 ? angle + 2*Math.PI : angle;
const angle = (vec) => posAngle(Math.atan2(vec[0], vec[1]));  // By switching x and y in the args to atan2, N = (0,1) becomes angle zero, and we get NESW sorting

const coordRange = Array.from({length:7}).map((_,n) => n - 3);
// The following sort yields the spiraling order O,N,E,S,W,NE,SE...
const spiralSortedCellVec = coordRange
                    .reduce ((a,y) => a.concat(coordRange.map (x => [x,y])), [])
                    .sort ((a, b) => taxicab(a) - taxicab(b)
                                  || maxDelta(a) - maxDelta(b)
                                  || angle(a) - angle(b));

let tmpCellIndex = coordRange.map(()=>coordRange.map(()=>null));
spiralSortedCellVec.forEach ((vec, idx) => tmpCellIndex[vec[0]+3][vec[1]+3] = idx);
const cellIndex = tmpCellIndex;
const lookupCellIndex = (vec) => cellIndex[(vec[0]+3+14)%7][(vec[1]+3+14)%7] | (maxDelta(vec) > 3 ? 128 : 0);
const xCoords = spiralSortedCellVec.map ((vec) => vec[0] + 3);
const yCoords = spiralSortedCellVec.map ((vec) => vec[1] + 3);

const rotate0 = (xy) => xy;
const rotate1 = (xy) => [xy[1],-xy[0]];
const rotate2 = (xy) => rotate1(rotate1(xy));
const rotate3 = (xy) => rotate1(rotate2(xy));
const rotations = [rotate0, rotate1, rotate2, rotate3];
const inverseRotations = [rotate0, rotate3, rotate2, rotate1];
const reflectX = (xy) => [xy[0],-xy[1]];
const reflectY = (xy) => rotate3(reflectX(rotate1(xy)));
const sumVec = (a, b) => a.map((ai,i) => ai + b[i]);
const translate = (xy1) => (xy2) => sumVec (xy1, xy2);
const makeTransformLookupTableRow = (f) => spiralSortedCellVec.map(f).map(lookupCellIndex);
const rotationLookupTable = rotations.map(makeTransformLookupTableRow);
const inverseRotationLookupTable = inverseRotations.map(makeTransformLookupTableRow);
const transformations = spiralSortedCellVec.map(translate).concat ([rotate1, rotate2, rotate3, reflectX, reflectY]);
const coordLookupTable = Array.from({length:64}).map((_,n) => (n % 8 == 7 || (n>>3) == 7) ? -1 : lookupCellIndex([(n%8)-3,(n>>3)-3]));
const transformLookupTable = transformations.map(makeTransformLookupTableRow).concat ([xCoords, yCoords, coordLookupTable]);

// Truly random access memory: the hookup to the larger storage is randomly translated and oriented,
// with connectivity persisting only until the next interrupt, at which a single undo is performed if
// and only if the software interrupt flag (I) is set at the time of the interrupt.
class BoardMemory {
    // Number of mapped cells for each neighborhood dimension
    static MAPPED_CELLS = { 2: 2, 3: 9, 5: 25, 7: 49 };

    constructor(seed = 42, boardSize = 256) {
        this._B = boardSize;
        this._N = 7;  // neighborhood dimension (2, 3, 5, or 7)
        this.storage = new Uint8Array (this.storageSize);
        this.mt = new MersenneTwister (seed);
        // Scheduler mode: 'random' (default) or 'checkerboard'
        this.schedulerMode = 'random';
        this._checkerboardIndex = 0;
        this._checkerboardPass = 0;
        this._checkerboardCells = null;  // lazily built
        this.sampleNextMove();
        this.resetUndoHistory();
        // Feature flags (set by controller from boardParams)
        this.orientedRegistersEnabled = true;
        this.lookupTablesEnabled = true;
    }

    get B() { return this._B }  // cells per dimension (X,Y)
    get M() { return 1024 }  // memory in bytes per cell
    get N() { return this._N }  // memory-mapped neighborhood dimension (2, 3, 5, or 7)
    get Nsquared() { return BoardMemory.MAPPED_CELLS[this._N] || 49 }  // mapped cell count
    get log2M() { return 10 }  // = log_2(M)
    get sqrtM() { return 32 }  // = sqrt(M)
    get storageSize() { return this.B * this.B * this.M; }  // 64Mb
    get neighborhoodSize() { return this.Nsquared * this.M; }  // mapped bytes
    get byteOffsetMask() { return this.M - 1 }  // 0x3FF

    // 1. Memory map - RAM
    // A. Used by OS, 0xF0 - 0xFF.
    // Rotated vectors. These are unmapped/mapped by the memory manager at each random orientation.
    get firstVectorAddr() { return 0x00F0 }
    get lastVectorAddr() { return 0x00F9 }  // note PCHI (0xf9) is a rotating address; its top bits are auto-rotated by the scheduler, allowing execution of code in neighboring cells

    // Controller-reserved vectors. Used to store registers between updates, and random numbers during updates
    get firstControllerAddr() { return 0x00FA }
    get lastControllerAddr() { return 0x00FF }

    // B. Reserved for visualizer, 0x3C0-0x3FF.
    // Conventionally, the upper 64 bytes are for visualization
    // The program is free to ignore these, they will not disrupt program flow
    // Debuggers should not rely on them!

    // 16x16 monochrome bitmap (32 bytes)
    get bitmapPixelsPerSide() { return 16 }  // bits per dimension (X,Y)
    get bitmapBytes() { return this.bitmapPixelsPerSide * this.bitmapPixelsPerSide / 8 }  // 32 bytes
    get bitmapAddr() { return 0x03C0 }       // 32-byte monochrome bitmap (1 bit/pixel)

    // Display name (28 bytes)
    get displayNameBytes() { return 28 }
    get displayNameAddr() { return 0x03E0 }

    // 0x03FC-0x03FE: reserved
    // Hue byte
    get hueAddr() { return 0x03FF }          // 1 byte: hue (0-255 maps to 0-360° HSV)

    // 2. Memory map - ROM
    // Lookup tables for common symmetry operations
    get firstLookupTableAddr() { return 0xE000 }
    get lastLookupTableAddr() { return 0xEFFF }

    // Full serialization & deserialization.
    // Storage is serialized as an array of byte values (not UTF-8 text,
    // which is lossy for arbitrary binary data — bytes 0x80-0xBF without
    // valid multi-byte leaders get replaced with U+FFFD on round-trip).
    get state() { return { storage: Array.from(this.storage),
                            iOrig: this.iOrig,
                            jOrig: this.jOrig,
                            orientation: this.orientation,
                            nextCycles: this.nextCycles,
                            mt: this.mt.mt,
                            mti: this.mt.mti,
                            schedulerMode: this.schedulerMode,
                            _checkerboardIndex: this._checkerboardIndex,
                            _checkerboardPass: this._checkerboardPass } }
    set state(s) {
        this.storage = new Uint8Array(Array.isArray(s.storage) ? s.storage : new TextEncoder().encode(s.storage));
        this.iOrig = s.iOrig;
        this.jOrig = s.jOrig;
        this.orientation = s.orientation;
        this.nextCycles = s.nextCycles;
        this.mt.mt = s.mt;
        this.mt.mti = s.mti;
        if (s.schedulerMode !== undefined) this.schedulerMode = s.schedulerMode;
        if (s._checkerboardIndex !== undefined) this._checkerboardIndex = s._checkerboardIndex;
        if (s._checkerboardPass !== undefined) this._checkerboardPass = s._checkerboardPass;
        if (this.schedulerMode === 'checkerboard') this._buildCheckerboardCells();
    }

    // Convert a neighborhood cell index (0-48) to a storage byte offset.
    // This is the base index into storage[] for that cell's first byte.
    // Precomputes the full address translation (spiral lookup, rotation,
    // coordinate wrapping) so callers can access storage[] directly.
    neighborCellStorageBase (neighIdx) {
        const [x, y] = spiralSortedCellVec[this.unrotate(neighIdx)];
        const i = this.wrapCoord (this.iOrig + x);
        const j = this.wrapCoord (this.jOrig + y);
        return this.M * this.ijToCellIndex(i, j);
    }

    // Accessors
    getByte (idx) { return this.storage[idx]; }
    setByteWithoutUndo (idx, val) { this.storage[idx] = val & 0xFF; }
    setByteWithUndo (idx, val) {
        if (this.undoHistory && !(idx in this.undoHistory))
            this.undoHistory[idx] = this.getByte(idx);
        this.setByteWithoutUndo (idx, val);
    }

    undoWrites() {
        Object.keys(this.undoHistory).forEach ((addr) => this.setByteWithoutUndo (parseInt(addr), this.undoHistory[addr]));
        this.resetUndoHistory();
    }

    resetUndoHistory() {
        this.undoHistory = {};
    }

    disableUndoHistory() {
        delete this.undoHistory;
    }

    ijToCellIndex (i, j) {
        return j + this.B * i;
    }

    ijbToByteIndex (i, j, b) {
        return this.M * this.ijToCellIndex(i,j) + b;
    }

    ijbFromByteIndex (byteIdx) {
        const b = byteIdx % this.M;
        const ij = Math.floor (byteIdx / this.M);
        const j = ij % this.B;
        const i = Math.floor (ij / this.B);
        return [i, j, b];
    }

    wrapCoord (k) { return (k + this.B) % this.B; }

    addrToCellCoords (addr) {
        const b = addr & this.byteOffsetMask;
        const [x, y] = spiralSortedCellVec[this.unrotate (addr >> this.log2M)];
        const i = this.wrapCoord (this.iOrig + x);
        const j = this.wrapCoord (this.jOrig + y);
        return [i, j, b];
    }

    addrToByteIndex (addr) {
        if (addr < 0 || addr >= this.neighborhoodSize)
            return -1;
        const [i, j, b] = this.addrToCellCoords (addr);
        return this.ijbToByteIndex (i, j, b);
    }

    addrIsInVectorRange (addr) {
        const b = addr & this.byteOffsetMask;
        return b >= this.firstVectorAddr && b <= this.lastVectorAddr;
    }

    valIsInVectorRange (val) {
        return ((val >> 2) & 0x3F) <= 48;
    }

    doRotateTopBits (addr, val) {
        return this.orientedRegistersEnabled && this.addrIsInVectorRange(addr) && this.valIsInVectorRange(val);
    }

    read (addr) {
        if (this.lookupTablesEnabled && addr >= this.firstLookupTableAddr && addr <= this.lastLookupTableAddr) {
            const nRow = (addr - this.firstLookupTableAddr) >> 6;
            const nCol = addr & 63;
            if (nRow < transformLookupTable.length && nCol < transformLookupTable[nRow].length)
                return transformLookupTable[nRow][nCol] & 0xFF;
            return 0;
        }
        const idx = this.addrToByteIndex (addr);
        const val = idx < 0 ? 0 : this.getByte (idx);
        return this.doRotateTopBits(addr,val) ? this.rotateTopBits(val) : val;
    }

    write (addr, val) {
        const idx = this.addrToByteIndex (addr);
        if (idx >= 0)
            this.setByteWithUndo (idx, this.doRotateTopBits(addr,val) ? this.unrotateTopBits(val) : val);
    }

    // rotation helpers
    rotate (n) {
        return rotationLookupTable[this.orientation][n];
    }

    unrotate (n) {
        return inverseRotationLookupTable[this.orientation][n];
    }

    rotateTopBits (val) {
        return (val & 3) | (this.rotate(val >> 2) << 2);
    }

    unrotateTopBits (val) {
        return (val & 3) | (this.unrotate(val >> 2) << 2);
    }

    // Build the checkerboard cell lists for the current board size.
    _buildCheckerboardCells() {
        const B = this.B;
        this._checkerboardCells = [[], []];  // [pass0 (even), pass1 (odd)]
        for (let i = 0; i < B; i++) {
            for (let j = 0; j < B; j++) {
                this._checkerboardCells[(i + j) % 2].push([i, j]);
            }
        }
    }

    // Sample the Poisson timer duration from the MT RNG.
    _sampleTimerAndRnd() {
        const rv2 = this.mt.int();  // log part of waiting time
        const rv3 = this.mt.real();  // fractional part of waiting time
        const rv4 = this.mt.int();  // stored in nextRnd
        const halfLife = 177;
        const cycleMultiplier = 16;
        let r = rv2;
        let nHalfLives = 0;
        while (nHalfLives < 32 && (r & 1)) {
            r = r >> 1;
            ++nHalfLives;
        }
        this.nextCycles = Math.ceil(cycleMultiplier * halfLife * (nHalfLives + rv3));
        this.nextRnd = rv4;
    }

    // randomly sample next move
    sampleNextMove() {
        if (this.schedulerMode === 'checkerboard') {
            // Checkerboard mode: deterministic cell order, random orientation and timer
            if (!this._checkerboardCells) this._buildCheckerboardCells();
            const pass = this._checkerboardPass;
            const idx = this._checkerboardIndex;
            const cell = this._checkerboardCells[pass][idx];
            this.iOrig = cell[0];
            this.jOrig = cell[1];
            // Still consume rv1 from MT to keep RNG consumption consistent
            const rv1 = this.mt.int();
            this.orientation = (rv1 >> 16) & 3;
            this._sampleTimerAndRnd();
            // Advance index
            this._checkerboardIndex++;
            if (this._checkerboardIndex >= this._checkerboardCells[pass].length) {
                this._checkerboardIndex = 0;
                this._checkerboardPass = 1 - this._checkerboardPass;
            }
        } else {
            // Random mode: original behavior
            const rv1 = this.mt.int();  // new origin and orientation
            this.iOrig = rv1 % this.B;
            this.jOrig = ((rv1 >> 8) & 0xFFFF) % this.B;
            this.orientation = (rv1 >> 16) & 3;
            this._sampleTimerAndRnd();
        }
    }
};


export { BoardMemory, cellIndex };

File: wasm/src/memory.rs
/// BoardMemory — manages the grid of interconnected 6502 cells.
/// Port of board/memory.js.

use crate::mt::MersenneTwister;
use crate::tables::{INV_ROTATION_TABLE, ROTATION_TABLE, SPIRAL, TRANSFORM_TABLE};
use std::collections::HashMap;

pub struct BoardMemory {
    pub storage: Vec<u8>,
    pub mt: MersenneTwister,
    pub board_size: usize,
    pub i_orig: usize,
    pub j_orig: usize,
    pub orientation: usize,
    pub next_cycles: u32,
    pub next_rnd: u32,
    undo_history: Option<HashMap<usize, u8>>,
}

// Constants
pub const M: usize = 1024; // bytes per cell
pub const LOG2_M: usize = 10;
pub const N: usize = 7; // neighborhood dimension
pub const N_SQUARED: usize = 49; // neighborhood cells
const BYTE_OFFSET_MASK: usize = M - 1; // 0x3FF

// Address ranges
const FIRST_VECTOR_ADDR: usize = 0x00F0;
const LAST_VECTOR_ADDR: usize = 0x00F9;
const FIRST_LOOKUP_TABLE_ADDR: usize = 0xE000;
const LAST_LOOKUP_TABLE_ADDR: usize = 0xEFFF;

impl BoardMemory {
    pub fn new(seed: u32, board_size: usize) -> Self {
        let storage_size = board_size * board_size * M;
        let mut mem = BoardMemory {
            storage: vec![0u8; storage_size],
            mt: MersenneTwister::new(seed),
            board_size,
            i_orig: 0,
            j_orig: 0,
            orientation: 0,
            next_cycles: 0,
            next_rnd: 0,
            undo_history: Some(HashMap::new()),
        };
        mem.sample_next_move();
        mem
    }

    #[inline]
    pub fn storage_size(&self) -> usize {
        self.board_size * self.board_size * M
    }

    #[inline]
    pub fn wrap_coord(&self, k: i32) -> usize {
        ((k + self.board_size as i32) as usize) % self.board_size
    }

    #[inline]
    pub fn ij_to_cell_index(&self, i: usize, j: usize) -> usize {
        j + self.board_size * i
    }

    #[inline]
    pub fn ijb_to_byte_index(&self, i: usize, j: usize, b: usize) -> usize {
        M * self.ij_to_cell_index(i, j) + b
    }

    pub fn ijb_from_byte_index(&self, byte_idx: usize) -> (usize, usize, usize) {
        let b = byte_idx % M;
        let ij = byte_idx / M;
        let j = ij % self.board_size;
        let i = ij / self.board_size;
        (i, j, b)
    }

    /// Convert a neighborhood cell index (0-48) to a storage byte offset (base address).
    pub fn neighbor_cell_storage_base(&self, neigh_idx: usize) -> usize {
        let unrotated = self.unrotate(neigh_idx);
        let spiral = &*SPIRAL;
        let (x, y) = spiral[unrotated];
        let i = self.wrap_coord(self.i_orig as i32 + x);
        let j = self.wrap_coord(self.j_orig as i32 + y);
        M * self.ij_to_cell_index(i, j)
    }

    // --- Byte access ---
    #[inline]
    pub fn get_byte(&self, idx: usize) -> u8 {
        self.storage[idx]
    }

    #[inline]
    pub fn set_byte_without_undo(&mut self, idx: usize, val: u8) {
        self.storage[idx] = val;
    }

    #[inline]
    pub fn set_byte_with_undo(&mut self, idx: usize, val: u8) {
        if let Some(ref mut history) = self.undo_history {
            history.entry(idx).or_insert_with(|| self.storage[idx]);
        }
        self.storage[idx] = val;
    }

    pub fn undo_writes(&mut self) {
        if let Some(history) = self.undo_history.take() {
            for (addr, val) in history {
                self.storage[addr] = val;
            }
        }
        self.undo_history = Some(HashMap::new());
    }

    pub fn reset_undo_history(&mut self) {
        self.undo_history = Some(HashMap::new());
    }

    pub fn disable_undo_history(&mut self) {
        self.undo_history = None;
    }

    /// Iterate over undo history keys (byte indices that were written).
    pub fn undo_history_keys(&self) -> Vec<usize> {
        match &self.undo_history {
            Some(h) => h.keys().copied().collect(),
            None => Vec::new(),
        }
    }

    // --- Rotation ---
    #[inline]
    pub fn rotate(&self, n: usize) -> usize {
        ROTATION_TABLE[self.orientation][n] as usize
    }

    #[inline]
    pub fn unrotate(&self, n: usize) -> usize {
        INV_ROTATION_TABLE[self.orientation][n] as usize
    }

    #[inline]
    fn rotate_top_bits(&self, val: u8) -> u8 {
        (val & 3) | ((self.rotate((val >> 2) as usize) as u8) << 2)
    }

    #[inline]
    fn unrotate_top_bits(&self, val: u8) -> u8 {
        (val & 3) | ((self.unrotate((val >> 2) as usize) as u8) << 2)
    }

    #[inline]
    fn addr_is_in_vector_range(addr: u16) -> bool {
        let b = (addr as usize) & BYTE_OFFSET_MASK;
        b >= FIRST_VECTOR_ADDR && b <= LAST_VECTOR_ADDR
    }

    #[inline]
    fn val_is_in_vector_range(val: u8) -> bool {
        ((val >> 2) & 0x3F) <= 48
    }

    #[inline]
    fn do_rotate_top_bits(addr: u16, val: u8) -> bool {
        Self::addr_is_in_vector_range(addr) && Self::val_is_in_vector_range(val)
    }

    // --- Address translation ---
    fn addr_to_byte_index(&self, addr: u16) -> Option<usize> {
        let addr = addr as usize;
        if addr >= N_SQUARED * M {
            return None;
        }
        let b = addr & BYTE_OFFSET_MASK;
        let cell_idx = addr >> LOG2_M;
        let spiral = &*SPIRAL;
        let unrotated = self.unrotate(cell_idx);
        if unrotated >= 49 {
            return None;
        }
        let (x, y) = spiral[unrotated];
        let i = self.wrap_coord(self.i_orig as i32 + x);
        let j = self.wrap_coord(self.j_orig as i32 + y);
        Some(self.ijb_to_byte_index(i, j, b))
    }

    // --- Memory-mapped read/write ---
    pub fn read(&self, addr: u16) -> u8 {
        let a = addr as usize;
        // ROM lookup tables
        if a >= FIRST_LOOKUP_TABLE_ADDR && a <= LAST_LOOKUP_TABLE_ADDR {
            let n_row = (a - FIRST_LOOKUP_TABLE_ADDR) >> 6;
            let n_col = a & 63;
            let table = &*TRANSFORM_TABLE;
            if n_row < table.len() && n_col < 64 {
                return table[n_row][n_col];
            }
            return 0;
        }
        // RAM
        match self.addr_to_byte_index(addr) {
            Some(idx) => {
                let val = self.storage[idx];
                if Self::do_rotate_top_bits(addr, val) {
                    self.rotate_top_bits(val)
                } else {
                    val
                }
            }
            None => 0,
        }
    }

    pub fn write(&mut self, addr: u16, val: u8) {
        if let Some(idx) = self.addr_to_byte_index(addr) {
            let stored = if Self::do_rotate_top_bits(addr, val) {
                self.unrotate_top_bits(val)
            } else {
                val
            };
            self.set_byte_with_undo(idx, stored);
        }
    }

    // --- Random sampling ---
    pub fn sample_next_move(&mut self) {
        let rv1 = self.mt.int();
        let rv2 = self.mt.int();
        let rv3 = self.mt.real();
        let rv4 = self.mt.int();

        self.i_orig = (rv1 as usize) % self.board_size;
        self.j_orig = (((rv1 >> 8) & 0xFFFF) as usize) % self.board_size;
        self.orientation = ((rv1 >> 16) & 3) as usize;

        let half_life: f64 = 177.0;
        let cycle_multiplier: f64 = 16.0;
        let mut r = rv2;
        let mut n_half_lives: u32 = 0;
        while n_half_lives < 32 && (r & 1) != 0 {
            r >>= 1;
            n_half_lives += 1;
        }
        self.next_cycles =
            (cycle_multiplier * half_life * (n_half_lives as f64 + rv3)).ceil() as u32;
        self.next_rnd = rv4;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_new_board_memory() {
        let mem = BoardMemory::new(42, 8);
        assert_eq!(mem.storage.len(), 8 * 8 * 1024);
        assert_eq!(mem.board_size, 8);
    }

    #[test]
    fn test_ij_roundtrip() {
        let mem = BoardMemory::new(42, 16);
        for i in 0..16 {
            for j in 0..16 {
                let idx = mem.ijb_to_byte_index(i, j, 0x42);
                let (ri, rj, rb) = mem.ijb_from_byte_index(idx);
                assert_eq!((ri, rj, rb), (i, j, 0x42));
            }
        }
    }

    #[test]
    fn test_wrap_coord() {
        let mem = BoardMemory::new(42, 8);
        assert_eq!(mem.wrap_coord(-1), 7);
        assert_eq!(mem.wrap_coord(8), 0);
        assert_eq!(mem.wrap_coord(3), 3);
    }

    #[test]
    fn test_undo_writes() {
        let mut mem = BoardMemory::new(42, 8);
        let idx = mem.ijb_to_byte_index(0, 0, 0);
        mem.storage[idx] = 0xAA;
        mem.reset_undo_history();
        mem.set_byte_with_undo(idx, 0xBB);
        assert_eq!(mem.storage[idx], 0xBB);
        mem.undo_writes();
        assert_eq!(mem.storage[idx], 0xAA);
    }

    #[test]
    fn test_read_write_ram() {
        let mut mem = BoardMemory::new(42, 8);
        // Write to address 0x0200 (cell 0, byte 0x200 — outside vector range)
        mem.write(0x0200, 0x42);
        assert_eq!(mem.read(0x0200), 0x42);
    }

    #[test]
    fn test_read_rom_lookup() {
        let mem = BoardMemory::new(42, 8);
        // ROM at 0xE000 should return transform table values
        let val = mem.read(0xE000);
        // First row, first col of transform table = translate by spiral[0]=(0,0) applied to spiral[0]=(0,0) = cell index 0
        assert_eq!(val, 0);
    }

    #[test]
    fn test_sample_next_move_deterministic() {
        let mut mem1 = BoardMemory::new(42, 256);
        let mut mem2 = BoardMemory::new(42, 256);
        // After construction, both should have sampled the same first move
        assert_eq!(mem1.i_orig, mem2.i_orig);
        assert_eq!(mem1.j_orig, mem2.j_orig);
        assert_eq!(mem1.orientation, mem2.orientation);
        assert_eq!(mem1.next_cycles, mem2.next_cycles);
        // Sample again
        mem1.sample_next_move();
        mem2.sample_next_move();
        assert_eq!(mem1.i_orig, mem2.i_orig);
        assert_eq!(mem1.next_cycles, mem2.next_cycles);
    }

    #[test]
    fn test_neighbor_cell_storage_base() {
        let mem = BoardMemory::new(42, 8);
        // Cell 0 should be the origin cell
        let base0 = mem.neighbor_cell_storage_base(0);
        let expected = M * mem.ij_to_cell_index(mem.i_orig, mem.j_orig);
        assert_eq!(base0, expected);
    }
}
